### PR TITLE
Updating `_get_noise_shape()` to work in graph mode

### PR DESCRIPTION
**Pull request by velocirabbit**
Using the `noise_shape` argument in the `tf.keras.layers.Dropout` initializer currently doesn't work with `tf.function` if you have variable batch sizes because TF can't get the correct dropout mask size to use each batch. This was due to using `x.shape.dims` to fill in batch-specific sizes, but that doesn't work when autographed because the values stay as `None`. Changing this to `tf.shape(x)` fixes that.

### CODE REVIEW

**Comment by rthadur** [comment]
@velocirabbit please sign CLA.

**Comment by velocirabbit** [comment]
> @velocirabbit please sign CLA.

Just signed it!

File: tensorflow/python/ops/nn_ops.py [hunk 1]
```diff
@@ -4012,9 +4012,9 @@ def _get_noise_shape(x, noise_shape):
 
   if x.shape.dims is not None and len(x.shape.dims) == len(noise_shape_.dims):
     new_dims = []
-    for i, dim in enumerate(x.shape.dims):
-      if noise_shape_.dims[i].value is None and dim.value is not None:
-        new_dims.append(dim.value)
+    for i, dim in enumerate(tf.shape(x)):
```
Comment (alextp) [inline tensorflow/python/ops/nn_ops.py]: Did you run this code at all? The call to enumerate() here would raise an exception in non-autograph graph mode, and the call to numpy() below will also fail in graph mode.

Comment (velocirabbit) [reply tensorflow/python/ops/nn_ops.py]: Whoops, you're right. I messed up the test implementation, and basically skipped over testing Dropout in any form completely. Made the fixes, and tested in both Graph and Eager modes.

**Review by alextp — CHANGES_REQUESTED** [review]
It looks a little better but still needs tests.

**Comment by rthadur** [comment]
Closing this PR as this not against `master`, please open a new PR against `master` 
CC @mihaimaruseac
